File: playwright-framework/pages/login-page.js
// pages/login-page.js

class LoginPage {
    /**
     * @param {import('@playwright/test').Page} page
     */
    constructor(page) {
        this.page = page;
        
        // 1. UI Selectors (Encapsulated Locators)
        this.usernameInput = page.locator('#username');
        this.passwordInput = page.locator('#password');
        this.loginButton = page.locator('button[id="submit-login"]');
        // Looks for either the success alert banner or the logout link button
        this.secureAreaHeader = page.locator('#flash');
    }

    // 2. Structural Interactivity Actions
    async enterUsername(username) {
        console.log('Entering username parameter credentials.');
        await this.usernameInput.fill(username);
    }

    async enterPassword(password) {
        console.log('Entering password security tokens securely.');
        await this.passwordInput.fill(password);
    }

    async clickLogin() {
        console.log('Submitting authentication execution forms.');
        await this.loginButton.click();
    }

    // High-level action wrapper flow
    async login(username, password) {
        await this.enterUsername(username);
        await this.enterPassword(password);
        await this.clickLogin();
    }

  async getFlashMessageText() {
        console.log('Retrieving the Successful message text ...');
        // Wait for the flash element to attach and pull its inner text
        const message = await this.page.locator('#flash').textContent();
        return message;
    }
}

module.exports = { LoginPage };

File: selenium-framework/src/main/java/pages/LoginPage.java
package pages;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import java.time.Duration;

public class LoginPage {
    
    private final WebDriver driver;
    private final WebDriverWait wait;
    private final Logger log = LogManager.getLogger(this.getClass());

    // 1. Private UI Selectors (Encapsulation)
    private final By Username = By.id("username");
    private final By passwordInput = By.id("password");
    private final By loginButton = By.xpath("//button[@id='submit-login']");
    private final By userProfileMenu = By.id("flash"); // Visible only when logged in

    // Constructor to pass the browser session downstream
    public LoginPage(WebDriver driver) {
        this.driver = driver;
        this.wait = new WebDriverWait(driver, Duration.ofSeconds(10));
    }

    // 2. Structural Interactivity Methods
    public void enterUsername(String username) {
        log.info("Entering username parameter credentials.");
        wait.until(ExpectedConditions.visibilityOfElementLocated(Username)).sendKeys(username);
    }

    public void enterPassword(String password) {
        log.info("Entering password security tokens securely.");
        wait.until(ExpectedConditions.visibilityOfElementLocated(passwordInput)).sendKeys(password);
    }

    public void clickLogin() {
        log.info("Submitting authentication execution forms.");
        wait.until(ExpectedConditions.elementToBeClickable(loginButton)).click();
    }

    // A clean wrapper method to handle the comprehensive action step
    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
    }

    public boolean isDashboardVisible() {
        try {
            return wait.until(ExpectedConditions.visibilityOfElementLocated(userProfileMenu)).isDisplayed();
        } catch (Exception e) {
            log.error("Login not successful: ", e);
            return false;
        }
    }
}